Mage/CommandCenter/TransportClient/CommandHTTPClient.cs

```diff
@@ -18,37 +30,49 @@ public class CommandHTTPClient : CommandTransportClient {
 		this.endpoint = baseUrl + "/" + appName;
 		this.headers = new Dictionary<string, string>(headers);
 	}
-	
+
 	//
-	public override void SendBatch(CommandBatch commandBatch) {
-		List<string> commands = new List<string> ();
+	public override void SerialiseBatch(CommandBatch commandBatch) {
+		List<string> commands = new List<string>();
 		List<string> data = new List<string>();
 
 		// Attach batch headers to post data
-		JArray batchHeader = new JArray();
-		string sessionKey = mage.session.GetSessionKey();
-		if (!string.IsNullOrEmpty(sessionKey)) {
-			JObject sessionHeader = new JObject();
-			sessionHeader.Add("name", new JValue("mage.session"));
-			sessionHeader.Add("key", new JValue(sessionKey));
-			batchHeader.Add(sessionHeader);
+		JArray batchHeaders = new JArray();
+		for (int batchHeaderI = 0; batchHeaderI < commandBatch.batchHeaders.Count; batchHeaderI += 1) {
+			Dictionary<string, string> batchHeader = commandBatch.batchHeaders[batchHeaderI];
+			batchHeaders.Add(JObject.FromObject(batchHeader));
 		}
-		data.Add(batchHeader.ToString(Newtonsoft.Json.Formatting.None));
+		data.Add(batchHeaders.ToString(Newtonsoft.Json.Formatting.None));
 
 		// Attach command names to url and parameters to post data
-		for (int batchId = 0; batchId < commandBatch.batchItems.Count; batchId += 1) {
-			CommandBatchItem commandItem = commandBatch.batchItems[batchId];
+		for (int batchItemI = 0; batchItemI < commandBatch.batchItems.Count; batchItemI += 1) {
+			CommandBatchItem commandItem = commandBatch.batchItems[batchItemI];
 			commands.Add(commandItem.commandName);
 			data.Add(commandItem.parameters.ToString(Newtonsoft.Json.Formatting.None));
 			logger.data(commandItem.parameters).verbose("sending command: " + commandItem.commandName);
 		}
 
-		// Serialise post data
 		string batchUrl = endpoint + "/" + String.Join(",", commands.ToArray()) + "?queryId=" + commandBatch.queryId.ToString();
 		string postData = string.Join("\n", data.ToArray());
-		
+
+		// Cached the serialisation
+		commandBatch.serialisedCache = (object)new CommandHttpClientCache(
+			batchUrl,
+			postData,
+			new Dictionary<string, string>(this.headers)
+		);
+	}
+	
+	//
+	public override void SendBatch(CommandBatch commandBatch) {
```

Comment:
Spoke with Almir, in this case the app should crash out, because it should never happen.